File: java/ql/test/utils/modelgenerator/dataflow/p/Sources.java
package p;

import java.io.IOException;
import java.io.InputStream;
import java.net.ServerSocket;
import java.net.URL;
import java.util.List;

public class Sources {

  // Defined as a source in the model file next to the test.
  // neutral=p;Sources;source;();summary;df-generated
  public String source() {
    return "";
  }

  // source=p;Sources;true;readUrl;(URL);;ReturnValue;remote;df-generated
  // sink=p;Sources;true;readUrl;(URL);;Argument[0];request-forgery;df-generated
  // neutral=p;Sources;readUrl;(URL);summary;df-generated
  public InputStream readUrl(final URL url) throws IOException {
    return url.openConnection().getInputStream();
  }

  // source=p;Sources;true;socketStream;();;ReturnValue;remote;df-generated
  // neutral=p;Sources;socketStream;();summary;df-generated
  public InputStream socketStream() throws IOException {
    ServerSocket socket = new ServerSocket(123);
    return socket.accept().getInputStream();
  }

  // source=p;Sources;true;wrappedSocketStream;();;ReturnValue;remote;df-generated
  // neutral=p;Sources;wrappedSocketStream;();summary;df-generated
  public InputStream wrappedSocketStream() throws IOException {
    return socketStream();
  }

  // source=p;Sources;true;sourceToParameter;(InputStream[],List);;Argument[0].ArrayElement;remote;df-generated
  // source=p;Sources;true;sourceToParameter;(InputStream[],List);;Argument[1].Element;remote;df-generated
  // neutral=p;Sources;sourceToParameter;(InputStream[],List);summary;df-generated
  public void sourceToParameter(InputStream[] streams, List<InputStream> otherStreams)
      throws IOException {
    ServerSocket socket = new ServerSocket(123);
    streams[0] = socket.accept().getInputStream();
    otherStreams.add(socket.accept().getInputStream());
  }

  // Not a new source because a simple type is used in an intermediate step
  // neutral=p;Sources;wrapSourceGetBool;();summary;df-generated
  public Boolean wrapSourceGetBool() {
    String s = source();
    return s == "hello";
  }

  public class SourceReader {
    @Override
    public String toString() {
      return source();
    }
  }

  public class MyContainer<T> {
    private T value;

    // neutral=p;Sources$MyContainer;read;();summary;df-generated
    public String read() {
      return value.toString();
    }
  }

  // Not a new source as this callable has been manually modelled
  // as source neutral.
  // neutral=p;Sources;manualNeutralSource;();summary;df-generated
  public String manualNeutralSource() {
    return source();
  }

  // Not a new source as this callable already has a manual source.
  // neutral=p;Sources;manualSourceAlreadyDefined;();summary;df-generated
  public String manualSourceAlreadyDefined() {
    return source();
  }
}

File: csharp/ql/test/utils/modelgenerator/dataflow/Sources.cs
using System;

namespace Sources;

public class NewSources
{
    // Defined as source in the extensions file next to the test.
    // neutral=Sources;NewSources;Source1;();summary;df-generated 
    public static string Source1() => throw null;

    // Defined as source in the extensions file next to the test.
    // neutral=Sources;NewSources;Source2;();summary;df-generated
    public static string Source2() => throw null;


    // New source
    // source=Sources;NewSources;false;WrapConsoleReadLine;();;ReturnValue;stdin;df-generated
    // neutral=Sources;NewSources;WrapConsoleReadLine;();summary;df-generated
    public string? WrapConsoleReadLine()
    {
        return Console.ReadLine();
    }

    // New source
    // source=Sources;NewSources;false;WrapConsoleReadLineAndProcees;(System.String);;ReturnValue;stdin;df-generated
    // neutral=Sources;NewSources;WrapConsoleReadLineAndProcees;(System.String);summary;df-generated
    public string WrapConsoleReadLineAndProcees(string prompt)
    {
        var s = Console.ReadLine();
        return string.IsNullOrEmpty(s) ? "" : s.ToUpper();
    }

    // NOT new source as method is private
    private string? PrivateWrapConsoleReadLine()
    {
        return Console.ReadLine();
    }

    // New source
    // source=Sources;NewSources;false;WrapConsoleReadKey;();;ReturnValue;stdin;df-generated
    // neutral=Sources;NewSources;WrapConsoleReadKey;();summary;df-generated
    public ConsoleKeyInfo WrapConsoleReadKey()
    {
        return Console.ReadKey();
    }

    // Not a new source because a simple type is used in an intermediate step
    // neutral=Sources;NewSources;WrapConsoleReadLineGetBool;();summary;df-generated
    public bool WrapConsoleReadLineGetBool()
    {
        var s = Console.ReadLine();
        return s == "hello";
    }

    public abstract class ValueReader
    {
        // neutral=Sources;NewSources+ValueReader;GetValue;();summary;df-generated
        public abstract string GetValue();
    }

    public class MyConsoleReader : ValueReader
    {
        // neutral=Sources;NewSources+MyConsoleReader;GetValue;();summary;df-generated
        public override string GetValue()
        {
            return Console.ReadLine();
        }
    }

    public class MyOtherReader : ValueReader
    {
        // neutral=Sources;NewSources+MyOtherReader;GetValue;();summary;df-generated
        public override string GetValue()
        {
            return "";
        }
    }

    public class MyContainer<T> where T : ValueReader
    {
        public T Value { get; set; }

        // neutral=Sources;NewSources+MyContainer<T>;Read;();summary;df-generated
        public string Read()
        {
            return Value.GetValue();
        }
    }

    // Not a new source as this callable has been manually modelled
    // as source neutral.
    // neutral=Sources;NewSources;ManualNeutralSource;();summary;df-generated
    public string ManualNeutralSource()
    {
        return Console.ReadLine();
    }

    // Not a new source as this callable already has a manual source.
    // neutral=Sources;NewSources;ManualSourceAlreadyDefined;();summary;df-generated
    public string ManualSourceAlreadyDefined()
    {
        return Console.ReadLine();
    }

    public abstract class DataReader
    {
        // neutral=Sources;NewSources+DataReader;Read;();summary;df-generated
        public abstract string Read();
    }

    public class DataReaderKind1 : DataReader
    {
        // neutral=Sources;NewSources+DataReaderKind1;Read;();summary;df-generated
        public override string Read()
        {
            return Source1();
        }
    }

    public sealed class DataReaderKind2 : DataReader
    {
        // neutral=Sources;NewSources+DataReaderKind2;Read;();summary;df-generated
        public override string Read()
        {
            return Source2();
        }
    }

    public class C1
    {
        // neutral=Sources;NewSources+C1;ToString;();summary;df-generated
        public override string ToString()
        {
            return Source1();
        }
    }

    public sealed class C2
    {
        // neutral=Sources;NewSources+C2;ToString;();summary;df-generated
        public override string ToString()
        {
            return Source1();
        }
    }

}
